File: chrome-extension/tab-follow.js
// Chrome can replace the tab under a page (prerender/instant activation, a
// discarded tab coming back): chrome.tabs.onReplaced reports the swap and the
// old id stops existing. An MCP session still holds the old id, so without a
// record every later command failed with "No tab with given id" while the page
// it was driving sat right there. The follower keeps exactly that record - and
// nothing looser: a tab the lost one opened, or one on the same URL, is not a
// replacement and is never followed, because this is the user's own browser.
// Commands addressed to a replaced tab reach its successor, the answer says so
// (tab_followed), and the server can ask `tabs.resolve` who took over.
const REPLACED_KEY = "wsn_replaced_tabs";
const LIMIT = 200;

export function createTabFollower(tabsApi, sessionStorage) {
  const replaced = new Map();   // removed id -> added id
  let restoring = null;

  // One shared promise: every caller waits for the same restore instead of the
  // second one racing ahead of a restore the first one started.
  function restore() {
    if (!restoring) {
      restoring = (async () => {
        try {
          const stored = (await sessionStorage?.get(REPLACED_KEY))?.[REPLACED_KEY];
          for (const [from, to] of Object.entries(stored || {})) {
            if (!replaced.has(Number(from))) replaced.set(Number(from), Number(to));
          }
        } catch (error) {
          console.warn("bridge: could not restore tab replacements", error);
        }
      })();
    }
    return restoring;
  }

  function persist() {
    const entries = [...replaced.entries()].slice(-LIMIT);
    replaced.clear();
    for (const [from, to] of entries) replaced.set(from, to);
    Promise.resolve(sessionStorage?.set({[REPLACED_KEY]: Object.fromEntries(entries)}))
      .catch(() => {});
  }

  function follow(tabId) {
    let current = Number(tabId);
    for (let hops = 0; hops < 16 && replaced.has(current); hops += 1) current = replaced.get(current);
    return current;
  }

  async function alive(tabId) {
    try {
      return Boolean(await tabsApi.get(Number(tabId)));
    } catch (error) {
      return false;
    }
  }

  tabsApi?.onReplaced?.addListener((addedTabId, removedTabId) => {
    replaced.set(Number(removedTabId), Number(addedTabId));
    persist();
  });

  return {
    // Rewrites a command's tabId to the tab that replaced it; returns the
    // params unchanged (same object) when there is no replacement on record.
    async redirect(params) {
      if (!params || params.tabId === undefined || params.tabId === null) return params;
      await restore();
      const followed = follow(params.tabId);
      return followed === Number(params.tabId) ? params : {...params, tabId: followed};
    },
    async resolve(tabId) {
      await restore();
      const requested = Number(tabId);
      const successor = follow(requested);
      const replacedBy = successor === requested ? null : successor;
      return {
        tabId: requested,
        alive: await alive(requested),
        replaced_by: replacedBy,
        successor_alive: replacedBy === null ? null : await alive(replacedBy),
      };
    },
  };
}

File: web_search_neo/cdp/tab_follow.py
"""Find the tab Chrome put in place of a session's vanished tab - and nothing else.

Chrome can swap the tab under a page without the page ever closing: a
prerendered or instant navigation replaces the tab and hands the page a new id
(``chrome.tabs.onReplaced``). The session still holds the old id, so its next
call fails with ``No tab with given id`` although its page is right there.

Only that record counts. The companion (1.18.4+) keeps it and answers
``tabs.resolve``; the successor must also still exist. There are deliberately no
heuristics - "a tab the lost one opened", "a tab on the same URL" - because this
is the user's own Chrome: a tab the agent's page opened (a payment window, an
OAuth popup) or one the user opened on the same address is not the agent's to
take over. No record, an older companion, or a companion that says "no
successor" all mean the same thing: there is nothing to follow.
"""
from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)


def _as_int(value: Any) -> int | None:
    if value is None or isinstance(value, bool):
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def find_successor(bridge: Any, tab_id: int | None) -> int | None:
    """The live tab Chrome recorded as replacing ``tab_id``, or ``None``."""
    if tab_id is None or bridge is None:
        return None
    try:
        answer = bridge.request("tabs.resolve", {"tabId": int(tab_id)}, timeout=3.0) or {}
    except Exception as exc:  # an older companion: "Unknown bridge method"
        logger.debug("tabs.resolve unavailable for tab %s: %s", tab_id, exc)
        return None
    if not isinstance(answer, dict):
        return None
    successor = _as_int(answer.get("replaced_by"))
    if successor is None or successor == int(tab_id) or answer.get("successor_alive") is not True:
        return None
    return successor
